app/models/manageiq/providers/openstack/inventory/parser/cloud_manager.rb

```diff
@@ -0,0 +1,458 @@
+class ManageIQ::Providers::Openstack::Inventory::Parser::CloudManager < ManagerRefresh::Inventory::Parser
+  include ManageIQ::Providers::Openstack::RefreshParserCommon::HelperMethods
+  include ManageIQ::Providers::Openstack::RefreshParserCommon::Images
+  include Vmdb::Logging
+
+  def parse
+    # the parser needs a reference to the os_handle
+    # for the logger in safe_call to work.
+    @os_handle = collector.os_handle
+
+    availability_zones
+    cloud_services
+    flavors
+    miq_templates
+    key_pairs
+    orchestration_stacks
+    quotas
+    vms
+    cloud_tenants
+    vnfs
+    vnfds
+  end
+
+  def cloud_services
+    related_infra_ems = collector.manager.provider.try(:infra_ems)
+    hosts = related_infra_ems.try(:hosts)
+
+    collector.cloud_services.each do |s|
+      host = hosts.try(:find) { |h| h.hypervisor_hostname == s.host.split('.').first }
+      system_services = host.try(:system_services)
+      system_service = system_services.try(:find) { |ss| ss.name =~ /#{s.binary}/ }
+
+      cloud_service = persister.cloud_services.find_or_build(s.id)
+      cloud_service[:ems_ref] = s.id
```

Comment:
you can define these keys in the defaults for the InventoryCollection as :inventory_object_attributes => [:ems_ref, :source, ..]

then you can use writers like

cloud_service.source = 'compute'

(:ems_ref should be already there using the find_or_build(s.id))